lib/runtime-core/src/parse.rs

```diff
@@ -183,45 +196,54 @@ pub fn read_module<
 
-                info.exports.insert(field.to_string(), export_index);
+                info.write()
+                    .unwrap()
+                    .exports
+                    .insert(field.to_string(), export_index);
             }
             ParserState::StartSectionEntry(start_index) => {
-                info.start_func = Some(FuncIndex::new(start_index as usize));
+                info.write().unwrap().start_func = Some(FuncIndex::new(start_index as usize));
             }
             ParserState::BeginFunctionBody { .. } => {
                 let id = func_count.wrapping_add(1);
                 func_count = id;
                 if func_count == 0 {
-                    info.namespace_table = namespace_builder.take().unwrap().finish();
-                    info.name_table = name_builder.take().unwrap().finish();
-                    mcg.feed_signatures(info.signatures.clone())
+                    info.write().unwrap().namespace_table =
+                        namespace_builder.take().expect("214").finish();
+                    info.write().unwrap().name_table = name_builder.take().expect("216").finish();
+                    mcg.feed_signatures(info.read().unwrap().signatures.clone())
                         .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
-                    mcg.feed_function_signatures(info.func_assoc.clone())
+                    mcg.feed_function_signatures(info.read().unwrap().func_assoc.clone())
                         .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
-                    mcg.check_precondition(&info)
+                    mcg.check_precondition(&info.read().unwrap())
                         .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
                 }
 
                 let fcg = mcg
-                    .next_function()
+                    .next_function(Arc::clone(&info))
                     .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
                 middlewares
                     .run(
                         Some(fcg),
                         Event::Internal(InternalEvent::FunctionBegin(id as u32)),
-                        &info,
+                        &info.read().unwrap(),
                     )
                     .map_err(|x| LoadError::Codegen(x))?;
 
-                let sig = info
+                let info_read = info.read().unwrap();
+                let sig = info_read
                     .signatures
                     .get(
                         *info
+                            .read()
+                            .unwrap()
                             .func_assoc
-                            .get(FuncIndex::new(id as usize + info.imported_functions.len()))
-                            .unwrap(),
+                            .get(FuncIndex::new(
+                                id as usize + info.read().unwrap().imported_functions.len(),
+                            ))
+                            .expect("242"),
```

Comment:
Should we change this number with something easier to debug?

Reply:
I'll change it back to unwrap as it was before, or I'll add an expect here in case we hit some potential bug.